python/ray/tune/ray_trial_executor.py

```diff
@@ -466,13 +467,41 @@ def save(self, trial, storage=Checkpoint.DISK):
         """Saves the trial's state to a checkpoint."""
         trial._checkpoint.storage = storage
         trial._checkpoint.last_result = trial.last_result
+
         if storage == Checkpoint.MEMORY:
             trial._checkpoint.value = trial.runner.save_to_object.remote()
         else:
-            with warn_if_slow("save_to_disk"):
-                trial._checkpoint.value = ray.get(trial.runner.save.remote())
+            # If enabled, saves best checkpoints into a best folder
+            if trial.keep_best_checkpoints_num and trial.results_since_checkpoint_cnt > 0:
+                current_attr_mean = trial.results_since_checkpoint_sum / trial.results_since_checkpoint_cnt
+                if trial.compare_checkpoints(current_attr_mean):
+                    trial.best_checkpoint_attr_value = current_attr_mean
+                    self._checkpoint_and_erase("best", trial)
+                trial.results_since_checkpoint_sum, trial.results_since_checkpoint_cnt = 0, 0
```

Comment:
Hm I found this tricky to understand. Is it that the best checkpoints accumulate if they exceed the mean, and we delete the old ones if they exceed limit?

Maybe this would simplify the accounting:
- for tracking, keep a dict of { checkpoint_path -> score value }

During checkpointing here:
- save the new checkpoint to disk and update the dict
- then call a erase() function which looks at the map, and deletes from disk the checkpoint with the worst score (perhaps always preserving the latest checkpoint for fault-recovery purposes)

I think that would also decouple the actual checkpoint vs erase code slightly more for easier readability.

Reply:
Current_attr_mean is calculated as the mean of attr between the last checkpoint and the current one. If it is greater than the last biggest current_attr_mean (trial.best_checkpoint_attr_value) the checkpoint will be saved in the best folder. So at any time you will have at most `--keep_best_checkpoints_num` checkpoints in the best folder.

Okay, but I don't really see how you don't need `--keep_best_checkpoints_num` since that tells you how many checkpoints is dict allowed to have. Otherwise its pretty much the same just using dict instead of a fifo list.

Reply:
I was thinking to apply just the keep num checkpoints limit here (no distinguishment between best vs time based), which would unify the best folder and the normal checkpoint dir.

Reply:
If I understand correctly, you propose that if the arg is specified --arg 20 we keep just the 20 best-rated checkpoints and nothing else, right?